wgsl/scanner.cc

```diff
@@ -0,0 +1,927 @@
+
+  /// Appends a bit to the back of the queue
+  void push_back(bool value) {
+    assert(count_ < CAPACITY_IN_BITS);
+    count_++;
+    (*this)[count_ - 1] = value;
+  }
+
+  /// @returns true if the queue holds no bits.
+  bool empty() const { return count_ == 0; }
+
+  /// @returns the number of bits held by the queue.
+  size_t count() const { return count_; }
+
+ private:
+  std::bitset<CAPACITY_IN_BITS> bits_;
+  size_t count_ = 0;        // number of bits contained
+  size_t read_offset_ = 0;  // read offset in bits
+};
+
+class Lexer {
+ public:
+  Lexer(TSLexer* l) : lexer_(l) {}
+
+  /// Advances the lexer by one code point.
+  void advance() { lexer_->advance(lexer_, /* whitespace */ false); }
+
+  /// Returns the next code point, advancing the lexer by one code point.
+  CodePoint next() {
+    CodePoint lookahead = lexer_->lookahead;
+    advance();
+    return lookahead;
+  }
+
+  /// @return the next code point without advancing the lexer, or kEOF if there
+  /// are no more code points
+  CodePoint peek() { return lexer_->eof(lexer_) ? kEOF : lexer_->lookahead; }
+
+  /// @return true if the next code point is equal to @p code_point.
+  /// @note if the code point was found, then the lexer is advanced to that code
+  /// point.
+  bool match(CodePoint code_point) {
+    if (peek() == code_point) {
+      advance();
+      return true;
+    }
+    return false;
+  }
+
+  /// @return true if the next code point is found in @p code_points.
+  /// @note if the code point was found, then the lexer is advanced to that code
+  /// point.
+  bool match_anyof(std::initializer_list<CodePoint> code_points) {
+    for (CodePoint code_point : code_points) {
+      if (match(code_point)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /// Attempts to match an identifier that starts with XIDStart and has any
+  /// number of XIDContinue code points.
+  bool match_identifier() {
+    if (!is_xid_start(peek())) {
+      return false;
+    }
+
+    std::stringstream ss;
+    bool is_ascii = true;
+    if (CodePoint start = next(); start < 0x80) {
+      ss.put(char(start));
+    } else {
+      is_ascii = false;
+    }
+
+    while (true) {
+      if (!is_xid_continue(peek())) {
+        break;
+      }
+      if (CodePoint code_point = next(); code_point < 0x80) {
+        ss.put(char(code_point));
+      } else {
+        is_ascii = false;
+      }
+    }
+
+    if (is_ascii) {
+      if (ss.str() == "var") {
+        LOG("var");
+        return false;
+      }
+      LOG("ident: '%s'", ss.str().c_str());
+    }
+
+    return true;
+  }
+
+  /// Attempts to match a /* block comment */
+  bool match_block_comment() {
+    if (!match('/') || !match('*')) {
+      return false;
+    }
+
+    size_t nesting = 1;
+    while (nesting > 0 && !match(kEOF)) {
+      if (match('/') && match('*')) {
+        nesting++;
+      } else if (match('*') && match('/')) {
+        nesting--;
+      } else {
+        next();
+      }
+    }
```

Comment:
This recognizes an unterminated block comment as a block comment. 
It should not.